### PR TITLE
fix: remove duplicated decodeURIComponent

### LINKED ISSUE #2725: Properly encode/decode the percentage character (% -> %25) across browsers and upon initial navigation
Currently, the percentage character creates multiple issues and inconsistencies. This is partly due to:

- the `%` is forbidden as part of the pathname in the URL per its spec (https://url.spec.whatwg.org/#path-state) if not escaped
- IE allows it because it follows a different spec (https://tools.ietf.org/html/rfc3986#section-2.4)

In Vue router, we have multiple inconsistencies between asterisk path, param paths and regular path regarding escaping these characters and whether the navigation is initial or triggered by a push/router-link

Given this set of routes in history mode

```js
[
  { path: '/foo/:p', component: Foo },
  {
    path: '*',
    component: Other,
  },
]
```

## History mode / Chrome, IE 11,10

| `to`       | route match | href produced in router-link | `$router` info after clicking | `$router` info on direct navigation         | remarks                                                                              |
| ---------- | ----------- | ---------------------------- | ----------------------------- | ------------------------------------------- | ------------------------------------------------------------------------------------ |
| `/%7D`     | `*`         | `/%7D`                       | `/%7D` - `/}`                 | `/}` - `/}`                                 | User should write `/}` as the link instead                                           |
| `/}`       | `*`         | `/%7D`                       | `/}` - `/}`                   | N/A (browser should encode it te be `/%7D`) | -                                                                                    |
| `/%25`     | `*`         | `/%25`                       | `/%25` - `/%`                 | ERROR                                       | Navigation should be possible. Should user also write `/%` to be consistent?         |
| `/%`       | `*`         | ERROR                        | N/A (not working)             | N/A (not valid)                             | Or should they escape the percent themselves because it's the only special character |
| `/foo/%7D` | `/foo/:p`   | `/%7D`                       | `/%7D` - p = `}`              | `/}` - p = `}`                              | User should write `/foo/}`                                                           |
| `/foo/}`   | `/foo/:p`   | `/%7D`                       | `/}` - p = `}`                | `/}` - p = `}`                              |
| `/foo/%25` | `/foo/:p`   | `/%25`                       | `/%25` - p = `%`              | ERROR                                       | Navigation should be possible                                                        |

## Query / TODO

---

Complete html file to test. Requires a `vue-router.js` file in the same folder

`index.html`
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>Document</title>
  </head>

  <body>
    <script src="https://unpkg.com/vue"></script>
    <!-- <script src="https://unpkg.com/vue-router"></script> -->
    <script src="/vue-router.js"></script>

    <div id="app">
      <pre>{{ $route.path }}</pre>
      <pre>{{ $route.params }}</pre>
      <router-link to="/">/home</router-link>
      <router-link to="/%7D">/%7D</router-link>
      <router-link to="/}">/}</router-link>
      <router-link to="/%25">/%25</router-link>
      <!-- <router-link to="/%">/%</router-link> -->
      <router-link to="/foo/%7D">/foo/%7D</router-link>
      <router-link to="/foo/}">/foo/}</router-link>
      <!-- <router-link to="/foo/%">/foo/%</router-link> -->
      <router-link to="/foo/%25">/foo/%25</router-link>
      <router-view></router-view>
    </div>

    <script>
      var Home = { template: '<div>Home</div>' }
      var Foo = { template: '<div>Foo: {{ $route.params.p }}</div>' }
      var Other = {
        template: '<div>Other: {{ $route.fullPath }}</div>',
      }
      var router = new VueRouter({
        mode: 'history',
        routes: [
          { path: '/', component: Home },
          { path: '/foo/:p', component: Foo },
          {
            path: '*',
            component: Other,
          },
        ],
      })

      new Vue({
        router: router,
        el: '#app',
      })
    </script>
  </body>
</html>
```

### LINKED ISSUE #3138: Double decoding of url params
### Version
3.1.6

### Reproduction link
[https://lyaskovec.github.io/tests/](https://lyaskovec.github.io/tests/)

### Steps to reproduce
There are two cases where you can see issue with double decoding of URL:
case 1:
- Go to the example page
- Click on link  /%25
- Click on link /home
- Try to use browsers navigation buttons

case2:
- Go to the example page
- Click on link  /%25
- reload the browser page

### What is expected?
It should show a correct page

### What is actually happening?
I got "URIError: URI malformed" error

### PR DESCRIPTION
**Pull request by mestevezdeister**
As it is mentioned by @claudiocastro I proposed the change made at #3268 adding some e2e tests, in order to get the fix merged. This change is a fix for the #2725 issue. 

Before this change the URL such as:
- /query/%25
- /query;q=%25

Produced an initial error: 
- URIError: URI malformed at decodeURIComponent (<anonymous>)

And recursive errors:
- TypeError: Cannot read property 'path' of undefined

Fix change fix the issue and other test cases still pass.

Best regards

Close #2725 
Close #3268 
Close #3138

### CODE REVIEW

File: test/e2e/specs/basic.js [hunk 1]
```diff
@@ -70,6 +71,15 @@ module.exports = {
       .assert.cssClassPresent('li:nth-child(8)', 'exact-active')
       .assert.attributeEquals('li:nth-child(8) a', 'class', '')
 
+      // encoded percentage as path param
+      // https://github.com/vuejs/vue-router/issues/2725
+      .url('http://localhost:8080/basic/query/A%25')
+      .waitForElementVisible('#app', 1000)
+      .assert.containsText('.view', 'query: A%')
```
Comment (posva) [inline test/e2e/specs/basic.js]: the test should exactly check the content or add some delimiters like "A%"

File: test/e2e/specs/basic.js [hunk 2]
```diff
@@ -70,6 +71,15 @@ module.exports = {
...
+      .click('li:nth-child(10) a')
+      .assert.urlEquals('http://localhost:8080/basic/query/A%25')
+      .assert.containsText('.view', 'query: A%')
```
Comment (posva) [inline test/e2e/specs/basic.js]: Same as above, this test is currently failing as the text shows `A%25` instead

**Review by posva — CHANGES_REQUESTED** [review]
Thanks for this! It still has a few breaking things that need to be fixed

**Comment by mestevezdeister** [comment]
I've added the restrictions you have proposed and solved the differences. The point was that when initially executes the application the route path was decoded (obtains the URL from the browser); whilst when changed the route from inside the application it was encoded (via clicking router-link). 

This fact is what initially produces the decoding error on initially load (issue #2725), because it tried to decode an already decoded URL.

I have assumed that the URL used as 'to' param in router-link must be decoded and it worked. As it can be checked in the test, when changing the route via clicking a router-link which has defined a decoded URL the application URL is properly encoded, so I assume it is right.

Thank you for your comments.

**Comment by posva** [comment]
> I have assumed that the URL used as 'to' param in router-link must be decoded and it worked. As it can be checked in the test, when changing the route via clicking a router-link which has defined a decoded URL the application URL is properly encoded, so I assume it is right.

That's the issue though, `to` as a string or `path` should be encoded 🤔 I will have to check this on older browsers first to make sure it still works

**Comment by mestevezdeister** [comment]
Ok @posva, thank you very much for you attention. Can I help in any way? or it is up to you checking the changes in old browsers?

File: examples/basic/app.js [hunk 3]
```diff
@@ -73,6 +75,7 @@ const vueInstance = new Vue({
           </li>
         </router-link>
         <li><router-link to="/foo" replace>/foo (replace)</router-link></li>
+        <li><router-link to="/query/A%">/query/A%</router-link></li>
```
Comment (posva) [inline examples/basic/app.js]: This has to be 

```suggestion
        <li><router-link to="/query/A%">/query/A%25</router-link></li>
```

in order to work in all browsers and when copying the URL into a new window to any browser after clicking the link

File: examples/hash-mode/app.js [hunk 4]
```diff
@@ -66,6 +68,7 @@ const vueInstance = new Vue({
         <li><router-link to="/é/ñ">/é/ñ</router-link></li>
         <li><router-link to="/é/ñ?t=%25ñ">/é/ñ?t=%ñ</router-link></li>
         <li><router-link to="/é/ñ#é">/é/ñ#é</router-link></li>
+        <li><router-link to="/query/A%">/query/A%</router-link></li>
```
Comment (posva) [inline examples/hash-mode/app.js]: same as above

**Review by posva — CHANGES_REQUESTED** [review]
These are the only changes in terms of tests that need to be done. The other assertions setup in e2e tests are correct

**Comment by mestevezdeister** [comment]
After doing some research, coding and testing I found out the following:

The URL got from window.location was decoded, as it can be seen in html5.js and hash.js. But, I cannot see where the URL got from the "to" attribute of router-link is decoded, neither on push, replace, ... methods. As posva said before, the mentioned attribute should be encoded, so I presume that the internal URL was different (encoded or not) depending on the way to access the route. If the application has initially load, the internal URL was decoded, but when change the route without reloading the page, the URL was encoded.

This point caused the duplication decodeURIComponent, which is the core of the issue.

In order to fix this I have made the following changes:
1. Remove the decoding of the initial URL in html5.js and hash.js, to ensure the internal URL is the same independently the way the route is accessed.
2. Apply de decode before executing route matching, instead of decoding each parameter independently, to ensure the application found the route.

With these changes all tests works. If I can help you in any way I would be glad to do it.

**Comment by posva** [comment]
Thanks a lot, I think this is starting to look good. Sorry, it's taking some time, I must make sure it doesn't break existing code. So far, it's working well on all my tests. One thing I need to add is the application not to crash when the URL is malformed (happens on old browsers like IE9 on `/oops%not-valid`). This is done by doing a `try catch` and using the original value if it fails.

**Comment by mestevezdeister** [comment]
I'm glad he have finally managed to fix the issue successfully! One more thing, what is the protocol or schedule to publish a new release? I am looking forward to finally installing it on my project and close the issues.

Many thanks

**Comment by posva** [comment]
I still need to fix a few introduced problems and will likely release it next week

**Comment by jaapio** [comment]
I just installed `dev` in our project to see the results of this patch. But it doesn't seem to solve the issue for us. We do have a URL like: `/records/$%25$%25$%25` I do see the `window.location` contains this exact value. 

The error: 
```
[Vue warn]: Error in beforeCreate hook: "URIError: malformed URI sequence"

(found in <Root>)
```

```
[Vue warn]: Error in render: "TypeError: route is undefined"

(found in <Root>)
```

This happens when a user refreshes the page.

**Comment by posva** [comment]
@jaapio you need to build vue-router to update the `dist` folder in order to use the patch

**Comment by jaapio** [comment]
Will check that

**Comment by ttshivers** [comment]
This change breaks existing behavior.

I have a route like `/join/:server` where server is a url encoded value like `https:%2F%2Fsubdomain.domain.com`.

Before v3.4.6, it decoded the server param to: `https://subdomain.domain.com`.
Now, in v.3.4.6, it does not: `https:%2F%2Fsubdomain.domain.com`.

**Comment by Dmitry-k42** [comment]
Still duplicating decodeURI in case of regex path. I found this code in function matchRoute:
```  let m
  try {
    m = decodeURI(path).match(regex)
  } catch (err) {
    if (process.env.NODE_ENV !== 'production') {
      warn(false, `Error decoding "${path}". Leaving it intact.`)
    }
  }

  if (!m) {
    return false
  } else if (!params) {
    return true
  }
```

When second decoding fails, function return false and route can not be matched. I think code must be edited like this:
```  let m
  try {
    path = decodeURI(path)
  } catch (err) {
    if (process.env.NODE_ENV !== 'production') {
      warn(false, `Error decoding "${path}". Leaving it intact.`)
    }
  }

  m = path.match(regex)

  if (!m) {
    return false
  } else if (!params) {
    return true
  }
```

**Comment by posva** [comment]
Thanks @Dmitry-k42

**Pull request merged**